timemachine/fe/reweighting.py

```diff
@@ -0,0 +1,192 @@
+__all__ = [
+    "one_sided_exp",
+    "construct_endpoint_reweighting_estimator",
+    "construct_mixture_reweighting_estimator",
+    "interpret_as_mixture_potential",
+]
+
+from typing import Callable, Collection
+
+from jax import numpy as np
+from jax.scipy.special import logsumexp
+
+Samples = Params = Collection
+Energies = Array = np.array
+BatchedReducedPotentialFxn = Callable[[Samples, Params], Energies]
+
+
+def log_mean(log_values: Array) -> float:
+    # log(mean(values))
+    # = log(sum(values) / len(values))
+    return logsumexp(log_values) - np.log(len(log_values))
+
+
+def estimate_log_z_ratio(log_importance_weights: Array) -> float:
+    # log(mean(importance_weights))
+    return log_mean(log_importance_weights)
+
+
+def one_sided_exp(delta_us: Array) -> float:
+    # delta_us = -log_importance_weights
+    # delta_f  = -log_z_ratio
+    return -estimate_log_z_ratio(-delta_us)
+
+
+def interpret_as_mixture_potential(u_kn: Array, f_k: Array, N_k: Array) -> Array:
+    """Interpret samples from multiple states as if they originate from a *single* state given by this potential.
+
+    Notes
+    -----
+    * This is nicely illustrated in ref [1] as a way to interpret MBAR [2]
+    * Depends on the accuracy of the input free energy estimates (f_k - f_k[0]) for the source states.
+    * This is not the only way to interpret samples from multiple distributions
+        as if they come from a single mixture distribution.
+        In ref [3], alternatives are systematically enumerated.
+        Assuming f_k are exact, the current approach corresponds to N3 in ref [3].
+        (N3 the most computationally expensive of the options in ref [3],
+        since it requires evaluating all k energy functions on every sample,
+        but it is also the lowest variance.)
+
+    References
+    ----------
+    [1] [Shirts, 2017] Reweighting from the mixture distribution as a better way to describe
+        the Multistate Bennett Acceptance Ratio
+        https://arxiv.org/abs/1704.00891
+    [2] [Shirts, Chodera, 2008] Statistically optimal analysis of samples from multiple equilibrium states.
+        J. Chem. Phys. 129:124105, 2008.
+        http://dx.doi.org/10.1063/1.2978177
+    [3] [Elvira+, 2019] Generalized multiple importance sampling
+        https://arxiv.org/abs/1511.03095
+    """
+    n_states, n_samples = u_kn.shape
+    N_k = np.array(N_k)
+    assert f_k.shape == (n_states,)
+    assert np.sum(N_k) == n_samples
+
+    mixture_u_n = -logsumexp(f_k - u_kn.T, b=N_k, axis=1)
+
+    assert mixture_u_n.shape == (n_samples,)
+
+    return mixture_u_n
+
+
+def construct_endpoint_reweighting_estimator(
+    samples_0: Samples,
+    samples_1: Samples,
+    batched_u_0_fxn: BatchedReducedPotentialFxn,
+    batched_u_1_fxn: BatchedReducedPotentialFxn,
```

Comment:
Naming -- I need a good variable name (and type annotation) for the user-provided function that computes a batch of energies.

Initially I had named this `vec_u_fxn`, but in discussion with @mcwitt noticed this is unwieldy (suggests vectorization, but this should be agnostic to how the batch is computed).

Currently using `batched_u_fxn: BatchedReducedPotentialFxn` (to be agnostic to implementation, but hint that we'll apply batch-specific performance optimizations). Not settled on this name, and the same concept will be used in a few other upcoming features.

@mcwitt has also suggested `map_` prefix.

Reply:
we use `batched_` pretty extensively in other parts of the code too